File: app/src/main/java/com/snaptiongame/app/data/providers/FriendProvider.java
package com.snaptiongame.app.data.providers;

import com.snaptiongame.app.data.api.SnaptionApi;
import com.snaptiongame.app.data.models.AddFriendRequest;
import com.snaptiongame.app.data.models.Friend;
import com.snaptiongame.app.data.providers.api.ApiProvider;

import java.util.List;

import io.reactivex.Completable;
import io.reactivex.Observable;
import io.reactivex.Single;

/**
 * @author Tyler Wong
 */

public class FriendProvider {
    private static SnaptionApi apiService = ApiProvider.getApiService();

    public static Observable<List<Friend>> getFriends(int page) {
        return apiService.getFriends(page);
    }

    public static Observable<List<Friend>> getFollowers() {
        return apiService.getFollowers();
    }

    public static Observable<List<Friend>> getFacebookFriends() {
        return apiService.getFacebookFriends();
    }

    public static Single<AddFriendRequest> addFriend(AddFriendRequest request) {
        return apiService.addFriend(request);
    }

    public static Completable removeFriend(AddFriendRequest request) {
        return apiService.deleteFriend(request);
    }

    public static Observable<List<Friend>> getAllFriends() {
        return Observable.range(1, Integer.MAX_VALUE)
                .concatMap(FriendProvider::getFriends)
                .takeWhile(friends -> !friends.isEmpty());
    }

    public static Single<Boolean> isFriend(int userId) {
        return getAllFriends()
                .flatMapIterable(friend -> friend)
                .map(friend -> friend.id)
                .contains(userId);
    }
}

File: app/src/main/java/com/snaptiongame/app/data/providers/FriendProvider.kt
@file:JvmName("FriendProvider")

package com.snaptiongame.app.data.providers

import com.snaptiongame.app.R.id.friends
import com.snaptiongame.app.data.models.AddFriendRequest
import com.snaptiongame.app.data.models.Friend
import com.snaptiongame.app.data.providers.api.ApiProvider

import io.reactivex.Completable
import io.reactivex.Observable
import io.reactivex.Single

/**
 * @author Tyler Wong
 */

private val apiService = ApiProvider.getApiService()

fun getFriends(page: Int): Observable<List<Friend>> = apiService.getFriends(page)

fun getFollowers(): Observable<List<Friend>> = apiService.getFollowers()

fun getFacebookFriends(): Observable<List<Friend>> = apiService.getFacebookFriends()

fun addFriend(request: AddFriendRequest): Single<AddFriendRequest> = apiService.addFriend(request)

fun removeFriend(request: AddFriendRequest): Completable = apiService.deleteFriend(request)

fun getAllFriends(): Observable<List<Friend>> =
        Observable.range(1, Integer.MAX_VALUE)
            .concatMap { getFriends(it) }
            .takeWhile { friends -> !friends.isEmpty() }

fun isFriend(userId: Int): Single<Boolean> =
        getAllFriends()
            .flatMapIterable { friend -> friend }
            .map { friend -> friend.id }
            .contains(userId)
